apps/prairielearn/src/server.js

```diff
@@ -857,6 +870,7 @@ export async function initExpress() {
   //////////////////////////////////////////////////////////////////////
   // API ///////////////////////////////////////////////////////////////
 
+  app.use('/pl/api/trpc', (await import('./api/trpc/index.js')).default);
```

Comment:
How do we deploy updates to tRPC routes? Some kind of multi-stage deploy where we do custom handling in each case? Is there anything clever to be done here, like adding a `version` parameter to each RPC call? It's certainly not clear to me that this helps at all.

On a related note, what do we do when tRPC itself updates? I gather it's backwards compatible if the server and client are on different minor version numbers, but I don't think there are any guarantees about major version bumps?

None of the above questions are meant to block this PR -- I'm just thinking out loud here.

Reply:
Very good questions.

Deploying a change to a route would indeed require multi-stage handling. I think that'd look more or less similar to database changes:

- For the addition of a parameter/property, we'd have to first deploy updated handlers that _optionally_ take the argument, then deploy code that actually uses them. We can then clean things up by making things non-optional in the types.
- For the deletion of a parameter/property, we'd first update the handler to allow it to be missing, then update all callsites to drop it, then finally remove it entirely from the handler.
- For the renaming of a parameter/property, we'd update the handler to allow either name, then update all callsites, then remove the old name from the handler.

This is of course fundamentally harder than traditional function calls we've always used, but I think this is more or less unavoidable when you cross a network boundary and can't guarantee versions on either side.

Something like a full blue/green deployment where we can guarantee that blue will only talk to blue and green would only talk to green could allow us to make changes without really thinking about things. Obviously we're not there yet and I'm not sure that complexity is better than just being careful with RPC changes.

Reply:
For major versions: I get the sense that the wire protocol is pretty stable (https://trpc.io/docs/rpc). If that changes... we'll cross that bridge when we come to it.

Reply:
This all makes a lot of sense, thanks! I agree that it's basically like a DB change and this complexity is inherent in RPC if you don't want to accept downtime (which we don't want).